File: backend/src/auth/auth.controller.ts
import { Body, Controller, Post, Req, HttpCode } from '@nestjs/common';
import { Throttle } from '@nestjs/throttler';
import { AuthService } from './auth.service';
import { LoginDto, VerifyOtpDto, RefreshDto, ForgotDto, ResetDto, ChangePasswordDto } from './dto';
import { Public, CurrentUser, AuthUser } from '../common/decorators';

const ctx = (req: any) => ({ ip: req.ip, userAgent: req.headers['user-agent'] });

@Controller('auth')
export class AuthController {
  constructor(private auth: AuthService) {}

  @Public() @Throttle({ default: { limit: 10, ttl: 60_000 } }) @Post('login') @HttpCode(200)
  login(@Body() d: LoginDto, @Req() r: any) { return this.auth.login(d.identifier, d.password, ctx(r)); }

  @Public() @Throttle({ default: { limit: 10, ttl: 60_000 } }) @Post('verify-otp') @HttpCode(200)
  verify(@Body() d: VerifyOtpDto, @Req() r: any) { return this.auth.verifyOtp(d.otpToken, d.code, ctx(r)); }

  @Public() @Post('refresh') @HttpCode(200)
  refresh(@Body() d: RefreshDto) { return this.auth.refresh(d.refreshToken); }

  @Post('logout') @HttpCode(200)
  logout(@CurrentUser() u: AuthUser, @Req() r: any) { return this.auth.logout(u.sub, ctx(r)); }

  @Public() @Throttle({ default: { limit: 5, ttl: 60_000 } }) @Post('forgot-password') @HttpCode(200)
  forgot(@Body() d: ForgotDto) { return this.auth.forgot(d.identifier); }

  @Public() @Post('reset-password') @HttpCode(200)
  reset(@Body() d: ResetDto, @Req() r: any) { return this.auth.reset(d.otpToken, d.code, d.newPassword, ctx(r)); }

  @Post('change-password') @HttpCode(200)
  change(@CurrentUser() u: AuthUser, @Body() d: ChangePasswordDto, @Req() r: any) { return this.auth.changePassword(u.sub, d.currentPassword, d.newPassword, ctx(r)); }
}

File: mobile/lib/features/auth/auth_controller.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import '../../core/api.dart';

class AuthUser {
  final String id, fullName, email, phone;
  final List<String> roles;
  AuthUser.fromJson(Map<String, dynamic> j)
      : id = j['id'], fullName = j['fullName'], email = j['email'], phone = j['phone'],
        roles = ((j['roles'] as List?) ?? []).map((r) => r is String ? r : r['role']['code'] as String).toList();
}

class AuthState {
  final bool ready;
  final AuthUser? user;
  const AuthState({this.ready = false, this.user});
}

class AuthController extends StateNotifier<AuthState> {
  AuthController(this.ref) : super(const AuthState()) { _restore(); }
  final Ref ref;
  get _dio => ref.read(dioProvider);
  get _storage => ref.read(storageProvider);

  Future<void> _restore() async {
    try {
      if (await _storage.read(key: 'access') != null) {
        final r = await _dio.get('/users/me');
        state = AuthState(ready: true, user: AuthUser.fromJson(r.data));
        return;
      }
    } catch (_) {}
    state = const AuthState(ready: true);
  }

  Future<String> login(String identifier, String password) async =>
      (await _dio.post('/auth/login', data: {'identifier': identifier, 'password': password})).data['otpToken'];

  Future<void> verifyOtp(String otpToken, String code) async {
    final r = await _dio.post('/auth/verify-otp', data: {'otpToken': otpToken, 'code': code});
    await _storage.write(key: 'access', value: r.data['accessToken']);
    await _storage.write(key: 'refresh', value: r.data['refreshToken']);
    state = AuthState(ready: true, user: AuthUser.fromJson(r.data['user']));
  }

  Future<String> forgot(String identifier) async =>
      (await _dio.post('/auth/forgot-password', data: {'identifier': identifier})).data['otpToken'];

  Future<void> reset(String otpToken, String code, String pw) =>
      _dio.post('/auth/reset-password', data: {'otpToken': otpToken, 'code': code, 'newPassword': pw});

  Future<void> changePassword(String current, String next) =>
      _dio.post('/auth/change-password', data: {'currentPassword': current, 'newPassword': next});

  Future<void> logout() async {
    try { await _dio.post('/auth/logout'); } catch (_) {}
    await _storage.deleteAll();
    state = const AuthState(ready: true);
  }
}

final authProvider = StateNotifierProvider<AuthController, AuthState>((ref) => AuthController(ref));
